### ISSUE TITLE
Idea for tracking userIds

### ISSUE THREAD

**Issue by jhuenges** [meteorbigdata/visit-tracker#3]
As mentioned in the [Todo](https://github.com/meteorbigdata/visit-tracker#todo), i would like to implement the tracking of the userId. 
My idea is ti make use of the Tracker.autorun function of Meteor and simply add another server method to add the userId once a user logged in. But I have one question: What should happen if there is more than one user who is using the same client? Should we simply store an array of userIds?

**Comment by andrewreedy** [comment]
@jhuenges we could possibly include userId in the returnVisits array and have it start a new visit if the userId is switched. What do you think?

**Comment by jhuenges** [comment]
Yeah, that could work. Would you propose to store an object?

**Comment by andrewreedy** [comment]
Yep an object. Just out of curiosity what type of project are you going to be using this on? I originally developed it for marketing analytics but it could be used for just general visit-tracking.

**Comment by jhuenges** [comment]
Currently I am not using it :D Just out of curiosity. But I guess one could use it not only for marketing but also for general tracking and analytics. What pages are most popular and so on. But I am most interested in knowing from where the visitors are and what they are doing.

**Comment by jhuenges** [comment]
Okay, short summary of what I have done so far:
- if a user is logged in and returns, the methode ```logReturnVisit``` will store the ```date``` and the```userId```
- if a user is _not_ logged in while returning to the site, ```logReturnVisit``` will only store the ```date```
- if a user is _not_ logged in but logs in later, the method ```logUserId``` is called and stores another ```returnVisit``` with a ```date``` and ```userId```
- if a visit is associated with a ```userId``` and another user logs in, the second ```userId``` is stored in a separat ```returnVisit``` entry (in the same ```visit```
- if, for some reason, a user is alreay logged in but has no stored visit, the method ```logVisit``` will add the ```returnVisit```-array and stores the initial visit as an return visit including the ```userId```

What do you think about that? Especially the last point...

**Comment by andrewreedy** [comment]
@jhuenges All of these additions sounds great! Good work adding persistent sessions too :+1:

**Comment by andrewreedy** [comment]
@jhuenges actually how do you feel about when the userId changes starting a new Visit instead of logging it as a return visit

**Comment by jhuenges** [comment]
Yeah, I guess we could do that :+1:

**Comment by jhuenges** [comment]
@andrewreedy I just thought about it. Wouldnt it be better to store the userId separated from the returnVisit? If we are going to create a new visit for each user,we could make the userId more accessible and store it on a higher level and not "hidden" in an array.

**Comment by andrewreedy** [comment]
Yea good call! I agree

**Comment by jhuenges** [comment]
Okay, so I made the change. But I stumbled upon a problem. Take the following szenarion:
- User 1 is logged in and has a ```currentVisit```
- User 1 logs out (```currentVisit``` is still stored)
- User 2 logs in on the same connection
-- the ```userId``` stored in ```currentVisit``` and the ```Meteor.userId()``` are not the same
--- create a new visit record
--- ```currentVisit``` now contains the visit for User 2
- User 2 logs out and User 1 logs in again
-- the ```userId``` stored in ```currentVisit``` and the ```Meteor.userId()``` are not the same
--- create a new visit record ...

So just because there are 2 users using the same computer you get multiple visits. 
Any suggestions? We could 'resume' the visit once we know the ```Meteor.userId()``` and store it as an ```returnVisit``` ...

**Comment by andrewreedy** [comment]
Good point. maybe I was wrong with the suggestion of always creating a new visit on login. A time based expiration on a visit may be a better solution. We should set a max time between return visits's, so that after X minutes/hours the return visit is considered a new visit. I think your idea is good for resuming the visit if its been under that X amount of time since their previous `returnVisit`.

**Comment by andrewreedy** [comment]
@jhuenges any luck with this?

**Comment by jhuenges** [comment]
I haven`t tried it yet - Sorry! It shouldn`t be that difficult. I ll try to do it in the next couple of days :)

**Comment by jhuenges** [comment]
@andrewreedy  any idea about how long the max time between visits should be? Or should it be configurable?

**Comment by andrewreedy** [comment]
@jhuenges Session variables aren't always tied to authentication. We can use localstorage (https://github.com/okgrow/meteor-persistent-session) Type #2
